src/ripple/rpc/handlers/DownloadShard.cpp

```diff
@@ -0,0 +1,320 @@
+                "an array"));
+    }
+
+    auto dCtx {std::make_shared<DownloadContext>()};
+    dCtx->shardStore = context.app.getShardStore();
+    dCtx->j = context.app.journal("DownloadShard");
+
+    // Validate shards
+    static const std::string ext {".tar.lz4"};
+    for (auto& it : context.params[jss::shards])
+    {
+        // Validate the index
+        if (!it.isMember(jss::index))
+            return RPC::missing_field_error(jss::index);
+        auto& jv {it[jss::index]};
+        if (!(jv.isUInt() || (jv.isInt() && jv.asInt() >= 0)))
+        {
+            return RPC::make_error(rpcINVALID_PARAMS,
+                RPC::expected_field_message(jss::index,
+                    "unsigned integer"));
+        }
+
+        // Validate the URL
+        if (!it.isMember(jss::url))
+            return RPC::missing_field_error(jss::url);
+        parsedURL url;
+        if (!parseUrl(url, it[jss::url].asString()) ||
+            url.scheme != "https" ||
+            url.domain.empty() ||
+            url.path.empty())
+        {
+            return RPC::invalid_field_error(jss::url);
+        }
+
+        // URL must point to tar archive using lz4 compression
+        auto archiveName {url.path.substr(url.path.find_last_of("/\\") + 1)};
+        if (archiveName.empty() || archiveName.size() <= ext.size())
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::url,
+                    "invalid archive name"));
+        }
+        if (!boost::iends_with(archiveName, ext))
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::url,
+                    "invalid archive extension"));
+        }
+
+        dCtx->shards.emplace(jv.asUInt(), std::move(url));
+    }
+
+    // Prepare the shard store
+    RangeSet<std::uint32_t> indexes;
+    for (auto const& shard : dCtx->shards)
+        indexes.add(shard.first);
+    indexes = dCtx->shardStore->prepareShard(indexes);
+    if (indexes.empty())
+        return RPC::make_param_error("invalid shard ids");
+
+    // Remove unsuitable shards
+    for (auto it = dCtx->shards.begin(); it != dCtx->shards.end();)
+    {
+        if (!boost::icl::contains(indexes, it->first))
+            it = dCtx->shards.erase(it);
+        else
+            ++it;
+    }
+
+    if (context.params.isMember(jss::validate))
+    {
+        if (!context.params[jss::validate].isBool())
+        {
+            dCtx->shardStore->removePreShard(indexes);
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::validate,
+                    "a bool"));
+        }
+        dCtx->validate = context.params[jss::validate].asBool();
+    }
+    if (dCtx->validate &&
+        context.app.getOPs().getOperatingMode() != NetworkOPs::omFULL)
```

Comment:
since validation happens sometime later (after download, unpack, and move), is the `omFULL` condition implicitly/explicitly checked again when the shard validation logic is called? To put it another way, what if the server falls out of sync between this point and when the shard is actually validated?

Reply:
The state is checked again when the validation occurs. If we've fallen out of sync at the time of validation, the import is skipped for the id. You bring up a good point. It's probably a good idea to defer each import/validation attempt if not synced. That would also remove the need to be synced when the RPC is received.